protocols/abci/addressing-1174.md

```diff
@@ -0,0 +1,230 @@
+
+An example of this is when `ProcessProposal` needs to take inputs from third-party entities
+(e.g. price oracles) that are not guaranteed to provide exactly the same values to
+different processes during the same height.
+Another example is when `ProcessProposal` needs to read the system clock in order to perform its checks
+(e.g. Proposer-Based Timestamp when expressed as an ABCI 1.0 application).
+
+In principle, if an application's implementation of `PrepareProposal` and `ProcessProposal`
+is not able to fulfill coherence and determinism requirements,
+CometBFT cannot guarantee consensus _termination_ in all runs of the system.
+As a result, the application designers must start considering both CometBFT and their application
+as one monolithic block, in order to reason about termination.
+We thus lose the modularity provided when fulfilling the ABCI 1.0 requirements.
+Remember that CometBFT's consensus algorithm (Tendermint) is a well-known algorithm that
+has been studied, reviewed, formally analyzed, model-checked, etc.
+The combination of CometBFT and an arbitrary application as one single algorithm cannot
+leverage that extensive body of research applied to the Tendermint algorithm.
+This situation is risky and undesirable.
+
+So, the questions that arise are the following.
+Can we come up with a set of weaker requirements
+that applications unable to fulfill the current ABCI 1.0 requirements
+can still fulfill?
+Is this set of weaker requirements still strong enough to guarantee consensus _termination_?
+
+## Solution Proposed
+
+### Modified consensus _validity_
+
+Function $valid(v, bc_{h-1})$, as explained above, exclusively depends on its inputs
+(a block, and the blockchain state at the previous height).
+So it is always supposed to provide the same result when called at the same height for the same inputs,
+no matter at which process.
+This was the main reason for introducing the determinism requirements on `ProcessProposal`
+in the ABCI 1.0 specification.
+
+If we are to relax the determinism requirements on the application,
+we first need to modify function $valid()$ to be of the form $valid(v, bc_{h-1}, x_p)$,
+where $x_p$ is a variable local to process $p$.
+As $x_p$ may be different from $x_q$ for two processes $p$ and $q$, using
+$valid(v, bc_{h-1}, x_p)$ to implement in `PrepareProposal`  and `ProcessProposal`
+may break Requirements 3 to 5.
+
+Consensus _validity_ property is then modified as follows:
+
+- _weak validity_: function $valid(v, bc_{h-1}, x_p)$ has returned _true_ at least once
+  by a correct process for the decided block $v$.
+
+### Eventual Requirements
+
+We now relax the relevant ABCI 1.0 requirements in the following way.
+
+* Requirement 3b [`PrepareProposal`, `ProcessProposal`, eventual coherence]:
+  There exists a round $r_s \ge 0$ of height $h$ such that,
+  for any two correct processes $p$ and $q$ and any round $r_p \geq r_s$,
```

Comment:
Weak Validity talks about one correct processes while Req. 3b. talks about all correct processes. I guess it would make sense to highlight that?

Reply:
Thanks, I'll mention it